diff --git a/api/src/main/java/net/luckperms/api/platform/Platform.java b/api/src/main/java/net/luckperms/api/platform/Platform.java
index 9ed38bee..5e64324e 100644
--- a/api/src/main/java/net/luckperms/api/platform/Platform.java
+++ b/api/src/main/java/net/luckperms/api/platform/Platform.java
@@ -75,7 +75,8 @@ public interface Platform {
         SPONGE("Sponge"),
         NUKKIT("Nukkit"),
         VELOCITY("Velocity"),
-        FABRIC("Fabric");
+        FABRIC("Fabric"),
+        MIRAI_CONSOLE("Mirai Console");
 
         private final String friendlyName;
 
diff --git a/common/src/main/java/me/lucko/luckperms/common/api/implementation/ApiPlatform.java b/common/src/main/java/me/lucko/luckperms/common/api/implementation/ApiPlatform.java
index d19884c7..1c5d918c 100644
--- a/common/src/main/java/me/lucko/luckperms/common/api/implementation/ApiPlatform.java
+++ b/common/src/main/java/me/lucko/luckperms/common/api/implementation/ApiPlatform.java
@@ -52,7 +52,8 @@ public class ApiPlatform implements Platform, PluginMetadata {
 
     @Override
     public @NonNull String getApiVersion() {
-        String[] version = this.plugin.getBootstrap().getVersion().split("\\.");
+        String[] version = this.plugin.getBootstrap().getVersionLuckPerms().split("\\.");
+        version[1] = version[1].replace("-SNAPSHOT", "");
         return version[0] + '.' + version[1];
     }
 
diff --git a/common/src/main/java/me/lucko/luckperms/common/command/CommandManager.java b/common/src/main/java/me/lucko/luckperms/common/command/CommandManager.java
index e9d20622..7df417f9 100644
--- a/common/src/main/java/me/lucko/luckperms/common/command/CommandManager.java
+++ b/common/src/main/java/me/lucko/luckperms/common/command/CommandManager.java
@@ -85,6 +85,7 @@ import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Function;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 /**
  * Root command manager for the '/luckperms' command.
@@ -104,7 +105,7 @@ public class CommandManager {
     public CommandManager(LuckPermsPlugin plugin) {
         this.plugin = plugin;
         this.tabCompletions = new TabCompletions(plugin);
-        this.mainCommands = ImmutableList.<Command<?>>builder()
+        this.mainCommands = Stream.concat(ImmutableList.<Command<?>>builder()
                 .add(new UserParentCommand())
                 .add(new GroupParentCommand())
                 .add(new TrackParentCommand())
@@ -130,14 +131,28 @@ public class CommandManager {
                 .add(new DeleteTrack())
                 .add(new ListTracks())
                 .build()
-                .stream()
+                .stream(),injectedCommands())
                 .collect(ImmutableCollectors.toMap(c -> c.getName().toLowerCase(Locale.ROOT), Function.identity()));
     }
 
+    protected Stream<Command<?>> injectedCommands() {
+        return Stream.empty();
+    }
+
     public LuckPermsPlugin getPlugin() {
         return this.plugin;
     }
 
+    protected Object preExecute(Sender sender, String label, List<String> arguments) {
+        return null;
+    }
+    protected boolean shouldRenderVersion(Sender sender, boolean hasPermAny, boolean isFirstTime) {
+        return true;
+    }
+    protected boolean shouldRendNoPermsForSubCommands(Sender sender) {
+        return true;
+    }
+
     public TabCompletions getTabCompletions() {
         return this.tabCompletions;
     }
@@ -208,10 +223,25 @@ public class CommandManager {
     }
 
     private void execute(Sender sender, String label, List<String> arguments) {
+        {
+            Object result = preExecute(sender, label, arguments);
+            if (result != null) return;
+        }
         applyConvenienceAliases(arguments, true);
 
         // Handle no arguments
         if (arguments.isEmpty() || arguments.size() == 1 && arguments.get(0).trim().isEmpty()) {
+            boolean lpm_hasAnyPermission = hasPermissionForAny(sender);
+            boolean lpm_isFirstTime = false;
+            if (!lpm_hasAnyPermission) {
+                Collection<? extends Group> groups = this.plugin.getGroupManager().getAll().values();
+                if (groups.size() <= 1 && groups.stream().allMatch(g -> g.normalData().isEmpty())) {
+                    lpm_isFirstTime = true;
+                }
+            }
+
+            if (shouldRenderVersion(sender, lpm_hasAnyPermission, lpm_isFirstTime)) {
+
             sender.sendMessage(Message.prefixed(Component.text()
                     .color(NamedTextColor.DARK_GREEN)
                     .append(Component.text("Running "))
@@ -221,16 +251,21 @@ public class CommandManager {
                     .append(Message.FULL_STOP)
             ));
 
-            if (hasPermissionForAny(sender)) {
+            }
+
+            if (lpm_hasAnyPermission) {
                 Message.VIEW_AVAILABLE_COMMANDS_PROMPT.send(sender, label);
                 return;
             }
 
-            Collection<? extends Group> groups = this.plugin.getGroupManager().getAll().values();
-            if (groups.size() <= 1 && groups.stream().allMatch(g -> g.normalData().isEmpty())) {
+            if (lpm_isFirstTime) {
                 Message.FIRST_TIME_SETUP.send(sender, label, sender.getName());
             } else {
+                if (shouldRendNoPermsForSubCommands(sender)) {
+
                 Message.NO_PERMISSION_FOR_SUBCOMMANDS.send(sender);
+
+                }
             }
             return;
         }
@@ -288,14 +323,6 @@ public class CommandManager {
     }
 
     private void sendCommandUsage(Sender sender, String label) {
-        sender.sendMessage(Message.prefixed(Component.text()
-                .color(NamedTextColor.DARK_GREEN)
-                .append(Component.text("Running "))
-                .append(Component.text(AbstractLuckPermsPlugin.getPluginName(), NamedTextColor.AQUA))
-                .append(Component.space())
-                .append(Component.text("v" + this.plugin.getBootstrap().getVersion(), NamedTextColor.AQUA))
-                .append(Message.FULL_STOP)
-        ));
 
         this.mainCommands.values().stream()
                 .filter(Command::shouldDisplay)
diff --git a/common/src/main/java/me/lucko/luckperms/common/command/access/CommandPermission.java b/common/src/main/java/me/lucko/luckperms/common/command/access/CommandPermission.java
index 606c4b79..b6f8a3d4 100644
--- a/common/src/main/java/me/lucko/luckperms/common/command/access/CommandPermission.java
+++ b/common/src/main/java/me/lucko/luckperms/common/command/access/CommandPermission.java
@@ -31,6 +31,7 @@ import me.lucko.luckperms.common.sender.Sender;
  * An enumeration of the permissions required to execute built in LuckPerms commands.
  */
 public enum CommandPermission {
+    LPM_DEBUG("lpmdebug", Type.NONE),
 
     SYNC("sync", Type.NONE),
     INFO("info", Type.NONE),
diff --git a/common/src/main/java/me/lucko/luckperms/common/command/spec/CommandSpec.java b/common/src/main/java/me/lucko/luckperms/common/command/spec/CommandSpec.java
index fe5a9740..2a1e7803 100644
--- a/common/src/main/java/me/lucko/luckperms/common/command/spec/CommandSpec.java
+++ b/common/src/main/java/me/lucko/luckperms/common/command/spec/CommandSpec.java
@@ -38,6 +38,7 @@ import java.util.Locale;
  */
 @SuppressWarnings("SpellCheckingInspection")
 public enum CommandSpec {
+    LPM_DEBUG("/%s lpmdebug"),
 
     USER("/%s user <user>"),
     GROUP("/%s group <group>"),
diff --git a/common/src/main/java/me/lucko/luckperms/common/commands/misc/EditorCommand.java b/common/src/main/java/me/lucko/luckperms/common/commands/misc/EditorCommand.java
index 459003b4..7821b69c 100644
--- a/common/src/main/java/me/lucko/luckperms/common/commands/misc/EditorCommand.java
+++ b/common/src/main/java/me/lucko/luckperms/common/commands/misc/EditorCommand.java
@@ -107,6 +107,8 @@ public class EditorCommand extends SingleCommand {
 
         Message.EDITOR_START.send(sender);
 
+        sender.flush();
+
         WebEditorRequest.generate(holders, tracks, sender, label, plugin)
                 .createSession(plugin, sender);
     }
diff --git a/common/src/main/java/me/lucko/luckperms/common/config/ConfigKeys.java b/common/src/main/java/me/lucko/luckperms/common/config/ConfigKeys.java
index 1aa8d32e..1e12221e 100644
--- a/common/src/main/java/me/lucko/luckperms/common/config/ConfigKeys.java
+++ b/common/src/main/java/me/lucko/luckperms/common/config/ConfigKeys.java
@@ -691,7 +691,7 @@ public final class ConfigKeys {
     /**
      * A list of the keys defined in this class.
      */
-    private static final List<SimpleConfigKey<?>> KEYS = KeyedConfiguration.initialise(ConfigKeys.class);
+    private static List<SimpleConfigKey<?>> KEYS = KeyedConfiguration.initialise(ConfigKeys.class);
 
     public static List<? extends ConfigKey<?>> getKeys() {
         return KEYS;
diff --git a/common/src/main/java/me/lucko/luckperms/common/dependencies/DependencyManager.java b/common/src/main/java/me/lucko/luckperms/common/dependencies/DependencyManager.java
index 696ec809..444f3d25 100644
--- a/common/src/main/java/me/lucko/luckperms/common/dependencies/DependencyManager.java
+++ b/common/src/main/java/me/lucko/luckperms/common/dependencies/DependencyManager.java
@@ -49,6 +49,7 @@ import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 /**
  * Loads and manages runtime dependencies for the plugin.
@@ -167,6 +168,7 @@ public class DependencyManager {
         // attempt to download the dependency from each repo in order.
         for (DependencyRepository repo : DependencyRepository.values()) {
             try {
+                plugin.getLogger().info("Downloading dependency: " + dependency + ", " + file.getFileName() + " from " + repo);
                 repo.download(dependency, file);
                 return file;
             } catch (DependencyDownloadException e) {
diff --git a/common/src/main/java/me/lucko/luckperms/common/dependencies/DependencyRepository.java b/common/src/main/java/me/lucko/luckperms/common/dependencies/DependencyRepository.java
index 3606a290..1c756db2 100644
--- a/common/src/main/java/me/lucko/luckperms/common/dependencies/DependencyRepository.java
+++ b/common/src/main/java/me/lucko/luckperms/common/dependencies/DependencyRepository.java
@@ -41,6 +41,16 @@ import java.util.concurrent.TimeUnit;
  */
 public enum DependencyRepository {
 
+    ALIYUN_MIRROR("https://maven.aliyun.com/repository/public") {
+        @Override
+        protected URLConnection openConnection(Dependency dependency) throws IOException {
+            URLConnection connection = super.openConnection(dependency);
+            connection.setConnectTimeout((int) TimeUnit.SECONDS.toMillis(5));
+            connection.setReadTimeout((int) TimeUnit.SECONDS.toMillis(10));
+            return connection;
+        }
+    },
+
     /**
      * Maven Central mirror repository.
      *
diff --git a/common/src/main/java/me/lucko/luckperms/common/locale/Message.java b/common/src/main/java/me/lucko/luckperms/common/locale/Message.java
index f6c51ae5..ddd8ab5c 100644
--- a/common/src/main/java/me/lucko/luckperms/common/locale/Message.java
+++ b/common/src/main/java/me/lucko/luckperms/common/locale/Message.java
@@ -1347,8 +1347,7 @@ public interface Message {
                     .append(text(':'))),
             prefixed(text()
                     .color(WHITE)
-                    .append(text("     "))
-                    .append(text(plugin.getBootstrap().getServerVersion()))),
+                    .append(text(plugin.getBootstrap().versionOnCommandRender()))),
             prefixed(text()
                     .color(AQUA)
                     .append(text("-  ", WHITE))
diff --git a/common/src/main/java/me/lucko/luckperms/common/plugin/bootstrap/LuckPermsBootstrap.java b/common/src/main/java/me/lucko/luckperms/common/plugin/bootstrap/LuckPermsBootstrap.java
index 2680ab00..c904d6fb 100644
--- a/common/src/main/java/me/lucko/luckperms/common/plugin/bootstrap/LuckPermsBootstrap.java
+++ b/common/src/main/java/me/lucko/luckperms/common/plugin/bootstrap/LuckPermsBootstrap.java
@@ -230,4 +230,14 @@ public interface LuckPermsBootstrap {
         return null;
     }
 
+
+    // LuckPerms Mirai - Start
+
+    default String versionOnCommandRender() {
+        return null;
+    }
+
+    String getVersionLuckPerms();
+
+    // LuckPerms Mirai - End
 }
diff --git a/common/src/main/java/me/lucko/luckperms/common/sender/Sender.java b/common/src/main/java/me/lucko/luckperms/common/sender/Sender.java
index 92d177e1..1edb628c 100644
--- a/common/src/main/java/me/lucko/luckperms/common/sender/Sender.java
+++ b/common/src/main/java/me/lucko/luckperms/common/sender/Sender.java
@@ -160,4 +160,12 @@ public interface Sender {
         return true;
     }
 
+    /**
+     * Flush cached message to real sender
+     *
+     * For LuckPerms-Mirai
+     */
+    default void flush() {}
+
+    default boolean isHoverEventSupported() { return false; }
 }
diff --git a/common/src/main/java/me/lucko/luckperms/common/verbose/VerboseListener.java b/common/src/main/java/me/lucko/luckperms/common/verbose/VerboseListener.java
index 8d942b90..dd70c6eb 100644
--- a/common/src/main/java/me/lucko/luckperms/common/verbose/VerboseListener.java
+++ b/common/src/main/java/me/lucko/luckperms/common/verbose/VerboseListener.java
@@ -150,7 +150,7 @@ public class VerboseListener {
     }
 
     private void sendNotification(VerboseEvent event) {
-        if (this.notifiedSender.isConsole()) {
+        if (!this.notifiedSender.isHoverEventSupported()) {
             // just send as a raw message
             if (event instanceof PermissionCheckEvent) {
                 PermissionCheckEvent permissionEvent = (PermissionCheckEvent) event;
diff --git a/settings.gradle b/settings.gradle
index 37f93867..00680e6e 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -14,15 +14,4 @@ include (
         'api',
         'common',
         'common:loader-utils',
-        'bukkit',
-        'bukkit:loader',
-        'bukkit-legacy',
-        'bukkit-legacy:loader',
-        'bungee',
-        'bungee:loader',
-        'fabric',
-        'nukkit',
-        'nukkit:loader',
-        'sponge', 'sponge:sponge-service', 'sponge:sponge-service-api6', 'sponge:sponge-service-api7',
-        'velocity'
 )
